src/crimson/net/Protocol.h

```diff
@@ -61,6 +57,17 @@ class Protocol {
   seastar::gate pending_dispatch;
   AuthConnectionMetaRef auth_meta;
 
+ private:
+  bool closed = false;
+  // become valid only after closed == true
+  seastar::shared_future<> close_ready;
+
+// TODO: encapsulate as a state-machine for write
```

Comment:
what is this commit for? could you move it into its own PR where the state machine is implemented?

Reply:
It is already a state-machine now, and I just group them together here.
https://github.com/ceph/ceph/pull/27788/commits/04969f81508238f5e0839ffbbb86dcbd98746d19 should be more clear.

I'm just wondering if we can make the `Protocol` class much more lightweight, containing only the protocol-level differences in handshake logic and buffer-layouts. Other parts like this can be moved out to `SocketConnection`, in the future.